### PR TITLE
fix: Support SQLAlchemy 1.4.X

### PR DESCRIPTION
**Pull request by dpgaspar**
### Description

This PR patches flask-sqlalchemy 2.X on FAB itself by using a custom `SignallingSession`.

The upstream patch request: https://github.com/pallets/flask-sqlalchemy/pull/1001

Also fixes FAB interface for checking if a join was already made since the previous approach was based on a private method that was removed on SQLAlchemy 1.4.X

@kaxil @ashb @betodealmeida @villebro 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Is CRUD MVC related.
- [ ] Is Auth, RBAC security related.
- [ ] Changes the security db schema.
- [ ] Introduces new feature
- [ ] Removes existing feature

### CODE REVIEW

**Review by ashb — APPROVED** [review]
LGTM form glancing at the code -- do you want us to test this change in Airflow before you merge and release this?

**Review by kaxil — APPROVED** [review]

File: setup.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def desc():
         "Flask-OpenID>=1.2.5, <2",
         # SQLAlchemy 1.4.0 breaks flask-sqlalchemy and sqlalchemy-utils
         "SQLAlchemy<1.4.0",
-        "Flask-SQLAlchemy>=2.4, <3",
+        "Flask-SQLAlchemy>=2.4, <5",
```
Comment (kaxil) [inline setup.py:58]: Should we unpin `"SQLAlchemy<1.4.0"` - L57 in this PR itself?

Comment (dpgaspar) [reply setup.py:58]: for sure, thats a mistake! thks

**Comment by dpgaspar** [comment]
> LGTM form glancing at the code -- do you want us to test this change in Airflow before you merge and release this?

That would be awesome, I'm making a release candidate soon also, maybe it's easier for you to test, but if any problems found I would have to revert/fix.

File: flask_appbuilder/models/sqla/__init__.py [hunk 2]
```diff
@@ -46,6 +82,17 @@ def get_tables_for_bind(self, bind=None):
                 result.append(tables[key])
         return result
 
+    def create_session(self, options):
```
Comment (dpgaspar) [inline flask_appbuilder/models/sqla/__init__.py:85]: untyped api upstream

File: flask_appbuilder/models/sqla/__init__.py [hunk 3]
```diff
@@ -21,6 +29,34 @@
 _camelcase_re = re.compile(r"([A-Z]+)(?=[a-z0-9])")
 
 
+class CustomSignallingSession(SignallingSession):
+    """
+    Custom Signaling Session to support SQLALchemy>=1.4 with flask-sqlalchemy 2.X
+    https://github.com/pallets/flask-sqlalchemy/issues/953
+    """
+
+    def get_bind(self, mapper=None, *args, **kwargs):
```
Comment (dpgaspar) [inline flask_appbuilder/models/sqla/__init__.py:38]: untyped api upstream

File: setup.py [hunk 4]
```diff
@@ -53,8 +53,6 @@ def desc():
         "Flask-Babel>=1, <3",
         "Flask-Login>=0.3, <0.5",
         "Flask-OpenID>=1.2.5, <2",
-        # SQLAlchemy 1.4.0 breaks flask-sqlalchemy and sqlalchemy-utils
-        "SQLAlchemy<1.4.0",
```
Comment (villebro) [inline setup.py:57]: nit: While I assume 1.4 will be followed by 2.0, maybe we should keep the restriction but move it one minor up, like
```
 "SQLAlchemy<1.5",
```

**Review by villebro — APPROVED** [review]
LGTM with minor nit

File: setup.py (hunk 4 above)
Comment (dpgaspar) [reply setup.py:57]: I understand your concern, mixed thoughts here. The cap responsibility should come from flask-sqlalchemy. But probably makes sense given FAB huge "dependency" on SQLAlchemy

**Pull request merged**